### PR TITLE
Queue Notifications

### PR DESCRIPTION
**Pull request by wkim10**
# Description

Added email notifications using SendGrid when a student is ready to be helped by TAs and for students at the top of the queue. The notifications are triggered whenever a TA clicks "Start helping" or "Mark as missing" in QuestionDetails.tsx.

The "ready to be helped" notification is triggered always when TA clicks "Start helping."

The other case (for a user being number 1 on the queue) was a little trickier. There is a chance it is triggered when a TA clicks either "Start helping" or "Mark as missing," as that is the only time when a student will possibly move up in the queue. If the current question that was "started being helped" or "marked as missing" is at the top of the queue, then a notification will be sent to the second person on the queue, as the second student will move up to become number 1 on the queue. In any other case (meaning current question is not at the top of the queue), then the question that is at the top of the queue will not change so no notification needs to be sent. Let me know if this logic is confusing or if there are any edge cases that I didn't catch with this.

Also if you need SendGrid key for .env let me know :)

## Issues

:)

## Screenshots

Screenshot 2025-01-20 at 5 00 54 PM
Screenshot 2025-01-20 at 5 01 14 PM

## Test

Need to do thorough testing for all possible edge cases. Try marking a student as missing and marking them as being helped. Try this with groups of students (meaning notification should be sent to each student in the group). Try marking a student in the middle of the queue as missing/being helped (which wouldn't trigger the notification for a student at the top of the queue because there was no change).

## Possible Downsides

Maybe there is a better way using cloud functions to ensure no edge cases are missing (we could constantly listen for the top of the queue and whenever there is a change, we just send an email to the new person at position 1). Also, I left some comments in GitHub in the code where there could be improvements in code structure, feel free to critique :)

## Additional Documentations

:)

### CODE REVIEW

File: src/app/components/board/QuestionDetails.tsx [hunk 1]
```diff
@@ -39,6 +43,10 @@ export const QuestionDetails = (props: QuestionDetailsProps) => {
   } = props;
 
   const user = getUserSessionOrRedirect();
+  const { questions } = useOfficeHour();
```
Comment (wkim10) [inline src/app/components/board/QuestionDetails.tsx]: Better way to do this? I didn't wanna get all the questions and do the filtering for pendingQuestions unless I really had to, but couldn't think of another way to handle the different cases (I'm only getting questions and pendingQuestions for use in sendTopUserNotif function)

File: src/app/components/board/QuestionDetails.tsx [hunk 2]
```diff
@@ -80,6 +89,21 @@ export const QuestionDetails = (props: QuestionDetailsProps) => {
     }
   };
 
+  const sendTopUserNotif = async () => {
+    if (pendingQuestions.length > 1 && pendingQuestions[0].id === question.id) {
```
Comment (wkim10) [inline src/app/components/board/QuestionDetails.tsx:134]: Let me know if this if statement logic is confusing. Idea is that you only wanna send a notification that a user is at the top of the queue if you are altering the question state for a question that is currently at the top of the queue (as that would change the question at the top of the queue to the question that is second on the queue). In any other case, the question at the top would not change, so no need to send this notification

File: src/app/components/board/QuestionDetails.tsx [hunk 3]
```diff
@@ -80,6 +89,21 @@ export const QuestionDetails = (props: QuestionDetailsProps) => {
...
+      pendingQuestions[1].group.forEach(async (userId) => {
+        const user = await getUser(userId);
+        if (user) {
+          await sendEmail({
```
Comment (locmai1) [inline src/app/components/board/QuestionDetails.tsx]: maybe we could have an emailTemplate that takes in a user email, this way we can also send custom notifs for missingPerson as well

File: src/app/components/board/QuestionDetails.tsx (hunk 1 above)
Comment (locmai1) [reply src/app/components/board/QuestionDetails.tsx]: Mmmm, I think we could make a new function in `client/questions.ts` where we make a batch call and we could pass in `courseId` and `batchSize` and that function could also sort by timestamps as well. Something along the lines of this:

```js
export const getBatchedQuestions = async (courseId: string, batchSize: number, stateFilter: string) => {
  const questionsQuery = query(
    collection(db, `courses/${courseId}/questions`).withConverter(questionConverter),
    where("state", "==", stateFilter),
    orderBy("timestamp"),
    limit(batchSize)
  );

  const snapshot = await getDocs(questionsQuery);
  const questionsDocs: IdentifiableQuestions = snapshot.docs.map((doc) => ({
    id: doc.id,
    ...doc.data(),
  }));

  return questionsDocs;
};
```

File: src/app/components/board/QuestionDetails.tsx (hunk 2 above)
Comment (locmai1) [reply src/app/components/board/QuestionDetails.tsx:134]: Makes sense to me 😼.

Adding onto it, is there any point that you'd think we have an overload of emails to send, which it might be beneficial to handle email sending through parallelization

File: src/app/components/board/QuestionDetails.tsx [hunk 4]
```diff
@@ -236,11 +260,19 @@ export const QuestionDetails = (props: QuestionDetailsProps) => {
...
                   await partialUpdateQuestion(question.id, courseId, {
                     state: QuestionState.IN_PROGRESS,
                     helpedBy: user.id,
                     helpedAt: serverTimestamp(),
                   });
+                  users.forEach(async (user) => {
+                    await sendEmail({
+                      email: user.email,
+                      subject: "The TAs are ready to help!",
+                      body: "The TAs are ready to help you now! Please listen for your name.",
```
Comment (lnguyen2693) [inline src/app/components/board/QuestionDetails.tsx]: Should we have a separate notification message for other people in the group? Like "The TAs are ready to help a question group that you join" or something.

**Review by lnguyen2693 — COMMENTED** [review]
Goat!! Love receiving emails from OfficeMins🤩
The deployment still failed though. Also the `QuestionDetails.tsx` becomes quite messy with the sending notification. Considering that we still have some other notification cases, I think it's better to separate notification stuff into a different file.

**Comment by locmai1** [comment]
> Goat!! Love receiving emails from OfficeMins🤩 The deployment still failed though. Also the `QuestionDetails.tsx` becomes quite messy with the sending notification. Considering that we still have some other notification cases, I think it's better to separate notification stuff into a different file.

@wkim10 @johnny-t06 could deployment failing be due to sendGrid API key not in prod `env`?

**Comment by wkim10** [comment]
> > Goat!! Love receiving emails from OfficeMins🤩 The deployment still failed though. Also the `QuestionDetails.tsx` becomes quite messy with the sending notification. Considering that we still have some other notification cases, I think it's better to separate notification stuff into a different file.
> 
> @wkim10 @johnny-t06 could deployment failing be due to sendGrid API key not in prod `env`?

Yeah Johnny and I checked and that's why deployment's failing

File: src/app/components/board/QuestionDetails.tsx (hunk 4 above)
Comment (wkim10) [reply src/app/components/board/QuestionDetails.tsx]: Ohhh interesting yeah that might be less confusing

File: src/app/components/board/QuestionDetails.tsx [hunk 5]
```diff
@@ -236,11 +266,27 @@ export const QuestionDetails = (props: QuestionDetailsProps) => {
...
                   });
+                  users.forEach(async (user, index) => {
+                    if (index === 0) {
+                      await sendEmail({
+                        email: user.email,
+                        subject: "The TAs are ready to help!",
+                        body: "The TAs are ready to help you now! Please listen for your name.",
+                      });
+                    } else {
+                      await sendEmail({
```
Comment (wkim10) [inline src/app/components/board/QuestionDetails.tsx]: This is separate email for other group members that Linh suggested, gonna need help testing since I need a group :)

File: src/app/services/client/question.ts [hunk 6]
```diff
@@ -141,3 +146,29 @@ export const deleteQuestion = async (courseId: string, questionId: string) => {
...
+
+  const snapshot = await getDocs(questionsQuery);
+
+  const questionsDocs: IdentifiableQuestions = snapshot.docs.map((doc) => ({
+    id: doc.id,
+    ...doc.data(),
+  }));
+
+  return questionsDocs;
+};
```
Comment (johnny-t06) [inline src/app/services/client/question.ts:190]: Very cool filtering Won!

**Review by johnny-t06 — APPROVED** [review]
LGTM Won!

File: src/app/services/client/question.ts (hunk 6 above)
Comment (wkim10) [reply src/app/services/client/question.ts:190]: Thank you Johnny!

File: src/app/components/board/QuestionDetails.tsx [hunk 7]
```diff
@@ -66,9 +70,13 @@ export const QuestionDetails = (props: QuestionDetailsProps) => {
     }
     setJoinGroup(!joinGroup);
   };
-  
+
   const onMissingRemove = async () => {
     if (question.state === QuestionState.PENDING) {
+      await sendTopUserNotif();
+      await sendEmail(
+        getEmailTemplate("STUDENT_MISSING", users[0]?.email ?? "")
```
Comment (johnny-t06) [inline src/app/components/board/QuestionDetails.tsx:80]: Added a STUDENT_MISSING email with a safe guard just in case since it depends on existence of the group.

**Review by johnny-t06 — APPROVED** [review]
LGTM, great job won!

File: src/app/components/board/QuestionDetails.tsx (hunk 7 above)
Comment (wkim10) [reply src/app/components/board/QuestionDetails.tsx:80]: Just tested this works great
